pkg/clusterapi/aws/utils.go

```diff
@@ -0,0 +1,157 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package aws
+
+import (
+	"fmt"
+	"sort"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	clusterv1 "sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/aws/aws-sdk-go/service/ec2/ec2iface"
+	"github.com/aws/aws-sdk-go/service/elb"
+	"github.com/aws/aws-sdk-go/service/elb/elbiface"
+
+	cov1 "github.com/openshift/cluster-operator/pkg/apis/clusteroperator/v1alpha1"
+)
+
+// SortInstances sorts the given slice of instances based on their launch time. The first item in the
+// slice after sorting will be the most recently launched, and can be considered the definitive instance
+// for the machine, a caller may wish to terminate all others. This function should only be called with
+// running instances, not those which are stopped or terminated.
+func SortInstances(instances []*ec2.Instance) {
+	sort.Slice(instances, func(i, j int) bool {
+		if instances[i].LaunchTime == nil && instances[j].LaunchTime == nil {
+			// No idea what to do here, should not be possible, just return the first.
+			return false
+		}
+		if instances[i].LaunchTime != nil && instances[j].LaunchTime == nil {
+			return true
+		}
+		if instances[i].LaunchTime == nil && instances[j].LaunchTime != nil {
+			return false
+		}
+		if (*instances[i].LaunchTime).After(*instances[j].LaunchTime) {
+			return true
+		}
+		return false
+	})
+}
+
+// GetInstance returns the AWS instance for a given machine. If multiple instances match our machine,
+// the most recently launched will be returned. If no instance exists, the exists bool will return false.
+func GetInstance(machine *clusterv1.Machine, client ec2iface.EC2API) (*ec2.Instance, bool, error) {
+	instances, err := GetRunningInstances(machine, client)
+	if err != nil {
+		return nil, false, err
+	}
+	if len(instances) == 0 {
+		return nil, false, nil
+	}
+
+	SortInstances(instances)
+	return instances[0], true, nil
+}
+
+// GetRunningInstances returns all running instances that have a tag matching our machine name,
+// and cluster ID.
+func GetRunningInstances(machine *clusterv1.Machine, client ec2iface.EC2API) ([]*ec2.Instance, error) {
+	instances := []*ec2.Instance{}
```

Comment:
Can we make a good guess on what the size of instances will be? Is there typically one instance per reservation, so that we could guess that the number of instances is equal to the number of reservations?

Reply:
A good guess would be one instance per reservation, I don't think anything we create would deviate from that, and I would *hope* that it's not possible to have a reservation return with no instances and we'd undershoot the size.

Reply:
Ah even then it's fine because it's a slice, will modify.